operations/mimir-mixin/dashboards/dashboard-queries.libsonnet

```diff
@@ -124,16 +175,26 @@
       labelValuesCardinalityQueriesPerSecond: queryPerSecond('labelValuesCardinality'),
 
       // Read failures rate as percentage of total requests.
-      readFailuresRate: |||
-        (
-            sum(rate(cortex_request_duration_seconds_count{%(queryFrontendMatcher)s, route=~"%(readHTTPRoutesRegex)s",status_code=~"5.*"}[$__rate_interval]))
-            or
-            # Handle the case no failure has been tracked yet.
-            vector(0)
-        )
-        /
-        sum(rate(cortex_request_duration_seconds_count{%(queryFrontendMatcher)s, route=~"%(readHTTPRoutesRegex)s"}[$__rate_interval]))
-      ||| % variables,
+      readFailuresRate:: {
+        local template = |||
+          (
+              sum(%(countFailQuery)s)
+              or
+              # Handle the case no failure has been tracked yet.
+              vector(0)
+          )
+          /
+          sum(%(countQuery)s)
+        |||,
+        classic: template % {
+          countFailQuery: utils.nativeClassicHistogramCountRate(p.readRequestsPerSecondMetric, p.readRequestsPerSecondSelector + ',status_code=~"5.*|error"').classic,
```

Comment:
Looking at the resulting dashboard and the code in jsonnet libs, this seems to run each query twice, once for native and once for classic, and then hide them based on < -Inf toggled by a multiplier. Then if you were to switch the toggle, it'd have to run both queries again and then hide the other one, and the queries seem different enough not to benefit from the cache. I don't know enough about PromQL or the dashboards to give a concrete suggestion, but is there no way to run only either the classic/native query based on the toggle?

Reply:
For example, since we are able to change the multiplier in the queries, can we change the entire query for the one we want to hide, e.g. `vector(0) < -1`?

Reply:
I don't expect people to switch many times. I did discuss alternate solution with Grafana folks, but there is no solution at the moment, closest feature request is https://github.com/grafana/grafana/issues/79848 which I commented on. We'll see if we can get some traction on that. Another idea would be to detect compare to `< -Inf` in the PromQL engine and shortcut the result as an optimization. Although I'd rather we solved it in Grafana.

Reply:
> For example, since we are able to change the multiplier in the queries, can we change the entire query for the one we want to hide, e.g. `vector(0) < -1`?

That was suggested by Grafana folks as well as alternate, but it seems very unfriendly to users that want to "Edit" or "Explore" a panel and drill down further. See https://community.grafana.com/t/can-i-use-variable-control-query-enable-disable/85409

Reply:
Given the limitations I guess this is the best we can do for now, we can figure out how to better optimise PromQL query execution in general separately later (: